Text pre-processing is the process of preparing text data so that machines can use the same to perform tasks like analysis, predictions, etc. There are many different steps in text pre-processing but in this article, we will only get familiar with stop words, why do we remove them, and the different libraries that can be used to remove them.



### What are stop words? 🤔
The words which are generally filtered out before processing a natural language are called stop words. These are actually the most common words in any language (like articles, prepositions, pronouns, conjunctions, etc) and does not add much information to the text. Examples of a few stop words in English are “the”, “a”, “an”, “so”, “what”.

### Why do we remove stop words? 🤷‍♀️
Stop words are available in abundance in any human language. By removing these words, we remove the low-level information from our text in order to give more focus to the important information. In order words, we can say that the removal of such words does not show any negative consequences on the model we train for our task.

Removal of stop words definitely reduces the dataset size and thus reduces the training time due to the fewer number of tokens involved in the training.



### Do we always remove stop words? Are they always useless for us? 🙋‍♀️

The answer is no! 🙅‍♂️

We do not always remove the stop words. The removal of stop words is highly dependent on the task we are performing and the goal we want to achieve. For example, if we are training a model that can perform the sentiment analysis task, we might not remove the stop words.

Movie review: “The movie was not good at all.”

Text after removal of stop words: “movie good”

We can clearly see that the review for the movie was negative. However, after the removal of stop words, the review became positive, which is not the reality. Thus, the removal of stop words can be problematic here.

Tasks like text classification do not generally need stop words as the other words present in the dataset are more important and give the general idea of the text. So, we generally remove stop words in such tasks.

#### Word of caution:
Before removing stop words, research a bit about your task and the problem you are trying to solve, and then make your decision.



### Natural Language Toolkit (NLTK)

NLTK is an amazing library to play with natural language. When you will start your NLP journey, this is the first library that you will use. The steps to import the library and the English stop words list is given below:



Documtation: https://pythonspot.com/nltk-stop-words/

In [1]:
import nltk
nltk.download('stopwords')


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [2]:
import nltk
from nltk.corpus import stopwords
sw_nltk = stopwords.words('english')


In [3]:
sw_nltk

['a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 "he'd",
 "he'll",
 'her',
 'here',
 'hers',
 'herself',
 "he's",
 'him',
 'himself',
 'his',
 'how',
 'i',
 "i'd",
 'if',
 "i'll",
 "i'm",
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it'd",
 "it'll",
 "it's",
 'its',
 'itself',
 "i've",
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'on

In [4]:
print(len(sw_nltk))

198


In [5]:
en_stops = set(stopwords.words('english'))
words = ['There', 'is', 'a', 'Deep','Learning','Course','in','CLoudyML']
for word in words:
    if word not in en_stops:
        print(word)

There
Deep
Learning
Course
CLoudyML


In [6]:
text = "When I first met her she was very quiet. She remained quiet during the entire two hour long journey from Stony Brook to New York."
words = [word for word in text.split() if word.lower() not in sw_nltk]
# print(words)
new_text = " ".join(words)
print(new_text)
print("Old length: ", len(text))
print("New length: ", len(new_text))

first met quiet. remained quiet entire two hour long journey Stony Brook New York.
Old length:  129
New length:  82


### spaCy

spaCy is an open-source software library for advanced NLP. This library is quite popular now and NLP practitioners use this to get their work done in the best way.

In [7]:
!pip install -U spacy

In [8]:
import spacy
#loading the english language small model of spacy
en = spacy.load('en_core_web_sm')
sw_spacy = en.Defaults.stop_words
print(sw_spacy)

{'my', 'must', 'say', 'became', 'four', 'into', 'often', 'is', 'although', 'wherein', 'mine', 'made', 'done', 'eleven', 'onto', 'have', 'whence', 'yours', 'very', '‘d', 'hundred', 'yourself', 'besides', '‘ve', 'hence', 'along', 'forty', 'but', 'more', "'ve", 'under', '‘re', '‘s', 'so', 'see', 'what', 'whenever', 'ours', 'becoming', 'be', 'your', "'re", 'would', 'using', 'behind', 'before', 'put', 'once', 'latterly', 'other', 'part', 'third', 'whom', 'hereupon', '’re', 'neither', 'thereupon', 'themselves', 'former', 'therein', 'front', 'several', 'am', 'indeed', 'too', '’ll', 'amount', 'still', 'beyond', 'yet', 'the', 'due', 'everything', 'others', 'something', 'namely', 'where', 'i', 'enough', 'latter', 'twenty', 'ca', 'none', '’d', 'up', 'an', 'yourselves', 'except', 'it', 'their', 'nothing', '’ve', 'off', 'anyhow', 'above', 'alone', 'from', 'just', 'both', 'after', 'around', 'again', 'afterwards', 'n’t', 'always', 'already', 'its', 'ten', 'about', 'last', 'these', 'doing', 'get', 'he

In [9]:
!python3 -m spacy download en_core_web_sm


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 81.9 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [ ]:
import spacy
#loading the english language small model of spacy
en = spacy.load('en_core_web_sm')
sw_spacy = en.Defaults.stop_words
print(sw_spacy)

In [10]:
print(len(sw_spacy))

326


In [11]:
text


'When I first met her she was very quiet. She remained quiet during the entire two hour long journey from Stony Brook to New York.'

In [12]:
words = [word for word in text.split() if word.lower() not in sw_spacy]
new_text = " ".join(words)
print(new_text)
print("Old length: ", len(text))
print("New length: ", len(new_text))


met quiet. remained quiet entire hour long journey Stony Brook New York.
Old length:  129
New length:  72


### Gensim:

Gensim (Generate Similar) is an open-source software library that uses modern statistical machine learning. According to Wikipedia, Gensim is designed to handle large text collections using data streaming and incremental online algorithms, which differentiates it from most other machine learning software packages that target only in-memory processing.

In [14]:
!pip install gensim
import gensim
from gensim.parsing.preprocessing import remove_stopwords, STOPWORDS
print(STOPWORDS)

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.9/27.9 MB 67.3 MB/s eta 0:00:00
frozenset({'my', 'fill', 'must', 'etc', 'say', 'became', 'four', 'into', 'often', 'is', 'although', 'wherein', 'mine', 'system', 'made', 'ltd', 'done', 'onto', 'eleven', 'interest', 'have', 'whence', 'yours', 'very', 'hundred', 'yourself', 'besides', 'hence', 'along', 'forty', 'but', 'more', 'amoungst', 'under', 'what', 'see', 'so', 'whenever', 'ours', 'becoming', 'be', 'your', 'con', 'would', 'sincere', 'using', 'behind', 'before', 'put', 'once', 'latterly', 'other', 'co', 'part', 'ie', 'third', 'whom', 'hereupon', 'neither', 'couldnt', 'thereupon', 'themselves', 'former', 'therein', 'front', 'de', 'cry', 'several', 'am', 'indeed', 'too', 'amount', 'still', 'beyond', 'yet', 'the', 'due', 'everything', 'kg', 'others', 'something', 'namely', 'where', 'i', 'enough', 'didn', 'latter', 'twenty', 'hasnt', 'none', 'mill', 'up', 'an', 'yourselves', 'except', 'it', 'their', 'nothing', 'off', 'anyhow', 'above', 'alon

In [15]:
print(len(STOPWORDS))

337


In [16]:
new_text = remove_stopwords(text)
print(new_text)
print("Old length: ", len(text))
print("New length: ", len(new_text))

When I met quiet. She remained quiet entire hour long journey Stony Brook New York.
Old length:  129
New length:  83


Removal of stop words reduced the length of the sentence from 129 to 83. We can see that even though the length of stop words in spaCy and Gensim is similar, the resulting text is quite different.



### Can I add my own stop words to the list? ✍️
Yes, we can also add custom stop words to the list of stop words available in these libraries to serve our purpose.

Here is the code to add some custom stop words to NLTK’s stop words list:



In [ ]:
text

In [17]:
sw_nltk.extend(['first', 'second', 'third', 'me'])

print(len(sw_nltk))

202


### Can I remove stop words from the premade list? 👋

Yes, if we want we can also remove stop words from the list available in these libraries.

Here is the code using the NLTK library:



In [ ]:
sw_nltk.remove('not')

In [ ]:
sw_nltk

['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

In [18]:
len(sw_nltk)

202

The stop word ‘not’ is now removed from the stop words list.



In [19]:
text

'When I first met her she was very quiet. She remained quiet during the entire two hour long journey from Stony Brook to New York.'

In [20]:
#create your custom stop words list
my_stop_words = ['her','me','i','she','it']
words = [word for word in text.split() if word.lower() not in my_stop_words]
new_text = " ".join(words)
print(new_text)

print("Old length: ", len(text))
print("New length: ", len(new_text))


When first met was very quiet. remained quiet during the entire two hour long journey from Stony Brook to New York.
Old length:  129
New length:  115


### Assignment:

In [22]:
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

text = "Hello everyone. Welcome to Intro to Machine Learning Applications. We are now learning important basics of NLP."

stop_words = set(stopwords.words('english'))
word_tokens = word_tokenize(text)

filtered_sentence = [w for w in word_tokens if not w in stop_words]

filtered_sentence = []

for w in word_tokens:
    if w not in stop_words:
        filtered_sentence.append(w)

print(word_tokens)
print(filtered_sentence)

text2 = ' '.join(filtered_sentence)

LookupError: 
**********************************************************************
  Resource punkt_tab not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('punkt_tab')
  
  For more information see: https://www.nltk.org/data.html

  Attempted to load tokenizers/punkt_tab/english/

  Searched in:
    - '/root/nltk_data'
    - '/usr/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/local/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/local/lib/nltk_data'
**********************************************************************


q2:

!wget https://www.dropbox.com/s/o8lxi6yrezmt5em/reviews.txt
   
1. Read a file from its URL
2. Extract the text and tokenize it meaningfully into words.
3. Print the entire text combined after tokenization.
4. Perform stemming using both porter and snowball stemmers. Which one works the best? Why?
5. Remove stopwords
